File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/data/PreviewTrialManager.kt
package com.jworks.kanjiquest.android.data

import android.content.Context
import com.jworks.kanjiquest.core.domain.model.GameMode
import java.time.LocalDate
import java.time.format.DateTimeFormatter

/**
 * Tracks daily preview trial usage for free-tier users.
 * Free users can try premium modes a limited number of times per day.
 */
class PreviewTrialManager(context: Context) {

    private val prefs = context.getSharedPreferences("kanjiquest_preview_trials", Context.MODE_PRIVATE)
    private val dateFormatter = DateTimeFormatter.ISO_LOCAL_DATE

    companion object {
        /** Max daily preview trials per mode */
        val TRIAL_LIMITS = mapOf(
            GameMode.WRITING to 3,
            GameMode.VOCABULARY to 3,
            GameMode.CAMERA_CHALLENGE to 1
        )
    }

    private fun todayKey(mode: GameMode): String {
        val today = LocalDate.now().format(dateFormatter)
        return "${mode.name}_$today"
    }

    /** Get remaining preview trials for a mode today */
    fun getRemainingTrials(mode: GameMode): Int {
        val limit = TRIAL_LIMITS[mode] ?: return 0
        val used = prefs.getInt(todayKey(mode), 0)
        return (limit - used).coerceAtLeast(0)
    }

    /** Use one preview trial. Returns true if trial was available and consumed. */
    fun usePreviewTrial(mode: GameMode): Boolean {
        val remaining = getRemainingTrials(mode)
        if (remaining <= 0) return false
        val key = todayKey(mode)
        val used = prefs.getInt(key, 0)
        prefs.edit().putInt(key, used + 1).apply()
        return true
    }

    /** Check if a mode has any remaining preview trials today */
    fun hasTrialsRemaining(mode: GameMode): Boolean {
        return getRemainingTrials(mode) > 0
    }

    /** Get total trial limit for a mode */
    fun getTrialLimit(mode: GameMode): Int {
        return TRIAL_LIMITS[mode] ?: 0
    }
}

File: ios-app/KanjiQuest/Data/PreviewTrialManager.swift
import Foundation
import SharedCore

/// Tracks daily preview trial usage for free-tier users.
/// Free users can try premium modes a limited number of times per day.
/// Swift equivalent of Android PreviewTrialManager (SharedPreferences → UserDefaults).
final class PreviewTrialManager {

    private let defaults = UserDefaults.standard
    private let prefix = "kanjiquest_preview_trials_"

    /// Max daily preview trials per mode
    static let trialLimits: [String: Int] = [
        "WRITING": 3,
        "VOCABULARY": 3,
        "CAMERA_CHALLENGE": 1
    ]

    private func todayKey(mode: String) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let today = formatter.string(from: Date())
        return "\(prefix)\(mode)_\(today)"
    }

    /// Get remaining preview trials for a mode today
    func getRemainingTrials(mode: String) -> Int {
        let limit = Self.trialLimits[mode] ?? 0
        if limit == 0 { return 0 }
        let used = defaults.integer(forKey: todayKey(mode: mode))
        return max(limit - used, 0)
    }

    /// Use one preview trial. Returns true if trial was available and consumed.
    func usePreviewTrial(mode: String) -> Bool {
        let remaining = getRemainingTrials(mode: mode)
        if remaining <= 0 { return false }
        let key = todayKey(mode: mode)
        let used = defaults.integer(forKey: key)
        defaults.set(used + 1, forKey: key)
        return true
    }

    /// Check if a mode has any remaining preview trials today
    func hasTrialsRemaining(mode: String) -> Bool {
        return getRemainingTrials(mode: mode) > 0
    }

    /// Get total trial limit for a mode
    func getTrialLimit(mode: String) -> Int {
        return Self.trialLimits[mode] ?? 0
    }

    /// Get preview trial info for display
    func getPreviewTrialInfo(mode: String) -> (remaining: Int, limit: Int) {
        let limit = getTrialLimit(mode: mode)
        let remaining = getRemainingTrials(mode: mode)
        return (remaining, limit)
    }
}
